dashboard.el

```diff
@@ -180,6 +163,78 @@ If MESSAGEBUF is not nil then MSG is also written in message buffer."
     (beginning-of-line)
     (widget-forward 1)))
 
+(defun dashboard-generate-banner ()
+  (when (goto-char (point-max))
+    (dashboard-insert-banner)
+    (dashboard-insert-page-break)))
+(defun dashboard-generate-recents ()
+  (recentf-mode)
+  (when (dashboard-insert-file-list
+	 "Recent Files:"
+	 (dashboard-subseq recentf-list 0 list-size))
+    (dashboard-insert--shortcut "r" "Recent Files:")
+    (dashboard-insert-page-break)))
+(defun dashboard-generate-bookmarks ()
+  (require 'bookmark)
+  (when (dashboard-insert-bookmark-list
+	 "Bookmarks:"
+	 (dashboard-subseq (bookmark-all-names)
+			   0 list-size))
+    (dashboard-insert--shortcut "m" "Bookmarks:")
+    (dashboard-insert-page-break)))
+(defun dashboard-generate-projects ()
+  (require 'projectile)
+  (projectile-mode)
+  (when (dashboard-insert-file-list
+	 "Projects:"
+	 (dashboard-subseq (projectile-relevant-known-projects)
+			   0 list-size))
+    (dashboard-insert--shortcut "p" "Projects:")
+    (dashboard-insert-page-break)))
+
+(defun dashboard-generate-fallback ()
```

Comment:
I dont like this fallback. Can we just do nothing if we encounter an invalid widget? or make emacs blow up and the user will figure out what is wrong